solana/solana-ibc/programs/solana-ibc/src/lib.rs

```diff
@@ -162,13 +161,78 @@ pub mod solana_ibc {
         ctx: Context<'a, 'a, 'a, 'info, Deliver<'info>>,
         message: ibc::MsgEnvelope,
     ) -> Result<()> {
+        if matches!(
+            message,
+            ibc::MsgEnvelope::Client(ibc::ClientMsg::UpdateClient(_))
+        ) {
+            panic!(
+                "Invalid message, Update Client message cannot be passed \
+                 here. Call send_update_client instead."
+            );
+        }
         let mut store = storage::from_ctx!(ctx, with accounts);
         let mut router = store.clone();
         ::ibc::core::entrypoint::dispatch(&mut store, &mut router, message)
             .map_err(error::Error::ContextError)
             .map_err(move |err| error!((&err)))
     }
 
+    /// Method to verify ed25519 signatures and send update client message.
+    ///
+    /// Since we cannot use ed25519 signature verification due to compute budget limit,
+    /// we have to call ed25519 program and pass the message, pubkey and signatures in
+    /// the same transaction in which the below method would be called. We then verify
+    /// the signatures comparing to the data in the previous instructions. If they
+    /// signature verification is success, we send the `UpdateClient` msg. This method
+    /// can only be called to send `UpdateClient` message. Calling otherwise would
+    /// panic.
+    pub fn send_update_client(
+        ctx: Context<SendUpdateClient>,
+        update_client_msg: ibc::MsgEnvelope,
+        signatures_data: Vec<SignatureData>,
+    ) -> Result<()> {
+        let length = signatures_data.len() as i64;
+        for (index, signature_data) in signatures_data.into_iter().enumerate() {
+            let verifier = ed25519::Verifier::new(
+                &ctx.accounts.ix_sysvar,
+                index as i64 - length,
+            )?;
```

Comment:
This is wrong. All signatures should be verified with a single call to the ed25519 program. Verifier is capable of handling that. You should initialise Verifier once and than call verify for each signature.

Reply:
oh i see, makes sense. Will do that.